cpp/src/quantiles/tdigest/tdigest.cu

```diff
@@ -0,0 +1,366 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/tdigest/tdigest.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/lists/lists_column_view.hpp>
+#include <cudf/structs/structs_column_view.hpp>
+#include <cudf/types.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+
+__device__ double lerp(double a, double b, double t) { return a + t * (b - a); }
+
+template <typename _T>
+struct cast_output {
+  data_type t;
+
+  template <typename T = _T, std::enable_if_t<!cudf::is_fixed_point<T>()>* = nullptr>
+  __device__ device_storage_type_t<T> operator()(double v)
```

Comment:
Are we able to merge this two functions together by implementing them with `if constexpr`?